nativelink-worker/src/directory_cache.rs

```diff
@@ -0,0 +1,535 @@
+            max_size_bytes: 10 * 1024 * 1024 * 1024, // 10 GB
+            cache_root: std::env::temp_dir().join("nativelink_directory_cache"),
+        }
+    }
+}
+
+/// Metadata for a cached directory
+#[derive(Debug, Clone)]
+struct CachedDirectoryMetadata {
+    /// Path to the cached directory
+    path: PathBuf,
+    /// Size in bytes
+    size: u64,
+    /// Last access time for LRU eviction
+    last_access: SystemTime,
+    /// Reference count (number of active users)
+    ref_count: usize,
+}
+
+/// High-performance directory cache that uses hardlinks to avoid repeated
+/// directory reconstruction from the CAS.
+///
+/// When actions need input directories, instead of fetching and reconstructing
+/// files from the CAS each time, we:
+/// 1. Check if we've already constructed this exact directory (by digest)
+/// 2. If yes, hardlink the entire tree to the action's workspace
+/// 3. If no, construct it once and cache for future use
+///
+/// This dramatically reduces I/O and improves action startup time.
+#[derive(Debug)]
+pub struct DirectoryCache {
+    /// Configuration
+    config: DirectoryCacheConfig,
+    /// Cache mapping digest -> metadata
+    cache: Arc<RwLock<HashMap<DigestInfo, CachedDirectoryMetadata>>>,
+    /// Lock for cache construction to prevent stampedes
+    construction_locks: Arc<Mutex<HashMap<DigestInfo, Arc<Mutex<()>>>>>,
+    /// CAS store for fetching directories
+    cas_store: Store,
+}
+
+impl DirectoryCache {
+    /// Creates a new DirectoryCache
+    pub async fn new(config: DirectoryCacheConfig, cas_store: Store) -> Result<Self, Error> {
+        // Ensure cache root exists
+        fs::create_dir_all(&config.cache_root)
+            .await
+            .err_tip(|| format!("Failed to create cache root: {:?}", config.cache_root))?;
+
+        Ok(Self {
+            config,
+            cache: Arc::new(RwLock::new(HashMap::new())),
+            construction_locks: Arc::new(Mutex::new(HashMap::new())),
+            cas_store,
+        })
+    }
+
+    /// Gets or creates a directory in the cache, then hardlinks it to the destination
+    ///
+    /// # Arguments
+    /// * `digest` - Digest of the root Directory proto
+    /// * `dest_path` - Where to hardlink/create the directory
+    ///
+    /// # Returns
+    /// * `Ok(true)` - Cache hit (directory was hardlinked)
+    /// * `Ok(false)` - Cache miss (directory was constructed)
+    /// * `Err` - Error during construction or hardlinking
+    pub async fn get_or_create(&self, digest: DigestInfo, dest_path: &Path) -> Result<bool, Error> {
+        // Fast path: check if already in cache
+        {
+            let mut cache = self.cache.write().await;
+            if let Some(metadata) = cache.get_mut(&digest) {
+                // Update access time and ref count
+                metadata.last_access = SystemTime::now();
+                metadata.ref_count += 1;
+
+                event!(
+                    Level::DEBUG,
```

Comment:
```suggestion
                debug!(
```
nit: Use the named helpers, rather than `Level`